### PR TITLE
Sort decorator

### PR DESCRIPTION
**Pull request by yuchaosydney**
Hi @NetanelBasal ,

Hope everything going good in your end.

This pull request is a property decorator which automatically sorts a property by a given property in each elements.

It basically uses setter accessor as a "watcher" of value change and sort the input value.

I have added unit tests which covers all senarios that I can think of including exception handling.

Appreciate if you could spend time on having a look and I am always happy to receive any feedback.

Greetings from yuchao :)

### CODE REVIEW

File: src/array-sort.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+export function arraySort<T>(sortByProperty: string, isDescending = false) {
+  const cachedValueKey = Symbol();
```
Comment (NetanelBasal) [inline src/array-sort.ts:2]: Sorry, I didn't saw this PR.

File: __tests__/decorators.spec.ts [hunk 2]
```diff
@@ -66,4 +66,110 @@ describe('Decorators', () => {
...
+        },
+        {
+          name: 'a'
+        }
+      ]);
+    });
+
+    it('should throw error if value is not an array', function() {
+      class TestTypeError {
+        @arraySort('name', true)
```
Comment (NetanelBasal) [inline __tests__/decorators.spec.ts:134]: Few improvements: 
1. Can we change it to `SortBy`?
2. Can we support types? `SortBy('name', { type: number/date/string })`
3. Can we support non-object items? `e.g: [3, 4, 6]`

File: src/array-sort.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+export function arraySort<T>(sortByProperty: string, isDescending = false) {
```
Comment (NetanelBasal) [inline src/array-sort.ts:1]: Change to `{ isDescending: true }`

File: src/array-sort.ts (hunk 1 above)
Comment (yuchaosydney) [reply src/array-sort.ts:2]: No worries at, I will address comments asap :)

File: __tests__/decorators.spec.ts (hunk 2 above)
Comment (yuchaosydney) [reply __tests__/decorators.spec.ts]: Hi @NetanelBasal ,

All three improvements have been addressed. Feel free to leave more feedback :)

Yuchao

File: src/array-sort.ts (hunk 3 above)
Comment (yuchaosydney) [reply src/array-sort.ts]: This is addressed

File: README.md [hunk 4]
```diff
@@ -118,6 +118,62 @@ export class AppComponent {
 }
 ```
 
+`arraySort` - sort an array by a specific property in individual elements
```
Comment (NetanelBasal) [inline README.md:121]: Update the docs, please.

Comment (yuchaosydney) [reply README.md]: Thanks for pointing it out, README is updated. :)

Yuchao

File: README.md [hunk 5]
```diff
@@ -118,6 +118,68 @@ export class AppComponent {
...
+      name: 'b'
+    },
+    {
+      name: 'a'
+    },
+    {
+      name: 'c'
+    }
+  ];
+  /* Output:
```
Comment (NetanelBasal) [inline README.md:142]: Can we make the readme cleaner, please? There is no need to show the output. And change everything to this style:
```ts
 @SortBy('name', {
    isDescending: false,
    type: 'string'
  })
  names = [{ name: 'a'}, { name: 'b' }]
```

Comment (yuchaosydney) [reply README.md]: Hi @NetanelBasal ,

Thanks for the feedback, it makes sense to keep README simple and clean.

README is updated and it shows how to use it and the decorator is capable of sort **date**, **string** & **number**. 

Please feel free to leave more feedback.

Greetings from Yuchao :)

File: src/sortby.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        if (!!sortByProperty && value.some(item => !item.hasOwnProperty(sortByProperty))) {
+          throw `Property ${propertyKey} has some elements that do not have property ${sortByProperty ? sortByProperty.toString() : ''}!`;
+        }
+      
+  
+        const errorMsg = (item: any, type: string) => `Value ${item} is not of predefined type ${type}`;
+
+        // Check if any element that does not follow pre-defined type
+        value.forEach((item: any) => {
+          switch(options.type) {
```
Comment (shaharkazaz) [inline src/sortby.ts:39]: Why validate the type? it's the consumer's responsibility

File: src/sortby.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    descriptor?: PropertyDescriptor
+  ) {
+    Object.defineProperty(target, propertyKey, {
+      set: function(value: Array<T>) {
+        
+        if (!value || !Array.isArray(value)) {
+          throw `Value of property ${propertyKey} is not a valid array!`;
+        }
+
+        if (!!sortByProperty && value.some(item => !item.hasOwnProperty(sortByProperty))) {
```
Comment (shaharkazaz) [inline src/sortby.ts:30]: I think we can skip this check if someone passed a `sortByProperty` it's the consumer's responsibility

File: README.md [hunk 8]
```diff
@@ -118,6 +118,31 @@ export class AppComponent {
 }
 ```
 
+`SortBy` - sort an array by a specific property in individual elements or non-object items
```
Comment (shaharkazaz) [inline README.md:121]: I think it's worth mentioning what is the default configuration 😄

File: src/sortby.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  type: 'string'
+}) {
+  const cachedValueKey = Symbol();
+  return function(
+    target: any,
+    propertyKey: string,
+    descriptor?: PropertyDescriptor
+  ) {
+    Object.defineProperty(target, propertyKey, {
+      set: function(value: Array<T>) {
```
Comment (shaharkazaz) [inline src/sortby.ts:24]: consider renaming `value` to `arr` or something that's indicating that this is the array just for clarity

File: src/sortby.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        if (!value || !Array.isArray(value)) {
+          throw `Value of property ${propertyKey} is not a valid array!`;
+        }
+
+        if (!!sortByProperty && value.some(item => !item.hasOwnProperty(sortByProperty))) {
+          throw `Property ${propertyKey} has some elements that do not have property ${sortByProperty ? sortByProperty.toString() : ''}!`;
+        }
+      
+  
+        const errorMsg = (item: any, type: string) => `Value ${item} is not of predefined type ${type}`;
```
Comment (shaharkazaz) [inline src/sortby.ts:35]: Move this function to the file level, it's static, no point of recreating it every time we use the decorator.

File: src/sortby.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+              if (sortByProperty ? !Date.parse(item[sortByProperty]) : !Date.parse(item)) {
+                throw errorMsg(item, options.type);
+              }
+              return; 
+            default:
+              throw errorMsg(item, options.type);
+          }
+        })
+
+        const sortFunc = (a: any, b: any) => {
```
Comment (shaharkazaz) [inline src/sortby.ts:60]: Move this function to the file level, it's static, no point of recreating it every time we use the decorator.
Also, you can apply the sorting direction while calling this function, if it's descending just multiply the result by `-1` then we save the call for reverse 😄 
What about nullish values? (if `a` or `b` are `undefined` or `null`) I would expect them to be pushed to the end.

File: src/sortby.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+          } else if (a > b) {
+            return 1;
+          }
+          return 0;
+        }
+       
+        // Perform sorting logic
+        if (sortByProperty) {
+          this[cachedValueKey] = value.sort(function(a: any, b: any) {
+            const aValue = options.type === 'date' ? new Date(a[sortByProperty]) : a[sortByProperty];
```
Comment (shaharkazaz) [inline src/sortby.ts:72]: cache `options.type === 'date'` in a variable.
Keep DRY, this section and the next are practically the same, merge them into one 👍

File: __tests__/decorators.spec.ts [hunk 13]
```diff
@@ -66,4 +66,177 @@ describe('Decorators', () => {
       expect(consoleSpy).toHaveBeenCalledTimes(1);
     });
   });
+
+  describe('arraySort', function() {
+
```
Comment (shaharkazaz) [inline __tests__/decorators.spec.ts:71]: DRY, you can use a factory function to generate all the test classes:
```
function testClass(array, config) => {
 return new class {
    @SortBy('name', config)
    testingArray = array
 }
}
```
Create a `describe` for each type you support and test both ascending and descending.
Also, test negative numbers and a `0` in the number type.

File: __tests__/decorators.spec.ts [hunk 14]
```diff
@@ -66,4 +66,177 @@ describe('Decorators', () => {
       expect(consoleSpy).toHaveBeenCalledTimes(1);
     });
   });
+
+  describe('arraySort', function() {
```
Comment (shaharkazaz) [inline __tests__/decorators.spec.ts:70]: Add a test for the default config

File: src/sortby.ts (hunk 6 above)
Comment (yuchaosydney) [reply src/sortby.ts]: Type check is removed.

File: src/sortby.ts (hunk 7 above)
Comment (yuchaosydney) [reply src/sortby.ts]: Also removed

File: README.md (hunk 8 above)
Comment (yuchaosydney) [reply README.md]: I agree, default configuration is written in README now

File: src/sortby.ts (hunk 9 above)
Comment (yuchaosydney) [reply src/sortby.ts]: Done

File: src/sortby.ts (hunk 10 above)
Comment (yuchaosydney) [reply src/sortby.ts]: I think it makes a lot of sense and this is my learning from this pull request. Thank you!

File: src/sortby.ts (hunk 11 above)
Comment (yuchaosydney) [reply src/sortby.ts]: Also makes a lot of sense, code is much cleaner after adopting above comment, really love it :)

In addition, nullish values are also considered and there is a test case which deliberately tests it.

File: src/sortby.ts (hunk 12 above)
Comment (yuchaosydney) [reply src/sortby.ts]: Like this comment, addressed in new commit.

File: __tests__/decorators.spec.ts (hunk 13 above)
Comment (yuchaosydney) [reply __tests__/decorators.spec.ts:86]: Love it as well, also addressed. New spec file looks much cleaner :)

File: __tests__/decorators.spec.ts (hunk 14 above)
Comment (yuchaosydney) [reply __tests__/decorators.spec.ts:70]: Addressed
